Miq expression removes sql only fields for filters

kbrock opened the pull request:
- [x] #22355 
- [x] #22356 

commits:

- added some comments around `MiqExpression`.
- added `prune_exp` to reduce the expression for a particular mode/language.
- using `prune_exp` for sql (was `preprocess_for_sql`)
- `Rbac#filter` now uses `prune_exp` to remove unnecessary sql from ruby.
- fix bugs around Demorgan's law (i.e.: `!(a || b) == !a && !b`)

code   | all(ms)  | records(ms) |do_eval(ms) | subst(ms) | comments
-------|-----:|--------:|-------:|------:|------
master | 2289 | 815     |    144 | 1294  |
after  | 1225 | 769     |     37 |  285  |
delta  |  46% | n/a     |    74% |  78%  |

Dates (via #22358) are on master and were a big win.
Since most of the fields were removed from the ruby expression processing, the whole processing time is approaching the record retrieval timing.

Wish I could have cached `@ruby` as something other than `true`, but since `YAML.parse` stores `@ruby==nil` as the uninitialized value, I need to store a value other than `nil` for a computed `nil` result.

NOTE: the expression was specifically picked because a lot of the fields were sql friendly. less sql friendly expressions will have less gain.

kbrock commented:
Update:
- rebased to take in external Count and dependent PR
- fixed whitespace issues and tweaked demorgan commit to remove duplicate code.

kbrock commented:
update:
- fixed refactor issue with renaming preprocess_exp!

kbrock commented:
update:
- added parameter to rbac to turn off prune_sql feature (mostly for performance testing)

kbrock commented:
update:
- Rbac#filtered takes a parameter to turn on/off the ruby filter

Fryguy commented on lib/miq_expression.rb:
```diff
@@ -300,42 +311,148 @@ def self._to_ruby(exp, context_type, tz)
+  end
+  private :reduce_exp_children
+
+  # private
```
Commented?

Fryguy commented on lib/miq_expression.rb:
```diff
@@ -323,31 +335,124 @@ def preprocess_exp!(exp)
+  end
+
+  # private
+  # When we hit a compount mode (OR, NOT, AND), we filter the individual children
```
```suggestion
  # When we hit a compound mode (OR, NOT, AND), we filter the individual children
```

Fryguy commented on lib/miq_expression.rb:
```diff
@@ -323,31 +335,124 @@ def preprocess_exp!(exp)
+  # Inside a NOT, the OR acts like the AND, and the AND acts like the OR
+  # so swap is introduced to handle Demorgan's law
+  #
+  def reduce_exp(exp, mode, swap: false)
```
Is reduce_exp modifying the original exp, or just returning a copy?   If the former, can we name it `reduce_exp!`

kbrock replied:
original expression is not changed, but it does point to some of the entries from it

jrafanie replied:
Sorry i'm late to this, but seems like we're dividing/filtering based on the mode (ruby/sql).  `reduce` makes me think of the ruby enumerable method and that we're just simplifying. Partitioning by mode is what I think this is doing after reviewing the code.  It's too bad partition is also an enumerable method.  I guess divide is the best I can offer.

kbrock replied:
I will rename to `prune_exp`. We are pruning out the nodes that are no relevant

kbrock commented:
update:
- changed wording in comments.
- added `@param` sections to a few methods
- cut down some wording on a few comments

Hope I didn't make comments less understandable when I cut them down.

kbrock commented:
@Fryguy are these comments easier to read now? Any last thoughts?

I'm pretty excited about the savings here. I hadn't expected the savings in `subst` to be the majority of our savings. It makes a bigger case for a ruby expression interpreter. This PR does bring us closer to implementing that as well.

Fryguy commented:
I've already approved 😄 - want a second review on this one - @jrafanie ?

jrafanie commented on lib/miq_expression.rb:
```diff
@@ -155,13 +161,23 @@ def self._to_human(exp, options = {})
+      pexp = preprocess_exp!(exp.deep_clone)
+      pexp, _ = reduce_exp(pexp, MODE_RUBY) if prune_sql
+      @ruby = self.class._to_ruby(pexp, context_type, timezone || "UTC".freeze) || true
+      @ruby == true ? nil : @ruby.dup
```
hmm, why was the `true` added?  Won't it be nil in that case anyway?

jrafanie replied:
The bigger concern is we're doubling down on caching options... timezone and now prune_sql. 😉

kbrock replied:
@jrafanie great question.

if `_to_ruby` is a no-op, it returns `nil`. This means the code should evaluate this to always be `true`. (You can see `lenient_evaluate` checking `nil?` first)

Unfortunately, `nil` is the same as undefined.

So I wanted to set a value that means "there is no ruby for us to do, just assume the logical value for this is true."

I tried changing the constructor to default `@ruby` to `false` (meaning it was undefined). But we deserialize these from `Yaml`, and yaml does not use the constructor but rather just defaults the values to `nil`.

kbrock replied:
> The bigger concern is we're doubling down on caching options... timezone and now prune_sql.

I did run into this issue when I was running benchmarks. I had to recreate the expression every time I changed the `prune_sql` value. I can't think of a reason why the code would pass this value.

I spent an hour looking for where we pass in timezones. The places that passed in a timezone derived the timezone from the report's time profile or the user. And those were only ever called from a single user.

I do remember working on code that populated reports that looked up all the timezones for the report group and generated report (values) for each timezone. I'm having trouble tracking this down though. But this is the only use case I can think of that would give the tz an issue.

Now as for the prune_sql. The only way this would be an issue is if we use a miqexpression in 2 places, one where we decide to use the expression to filter in sql and one where we don't. Feels very far fetched, but I guess it is possible.

jrafanie commented on lib/miq_expression.rb:
```diff
@@ -323,31 +340,121 @@ def preprocess_exp!(exp)
     operator = exp.keys.first
-    case operator.downcase
-    when "and"
-      exp[operator].dup.each { |atom| preprocess_for_sql(atom, attrs) }
```
Why was this dup here?  Do we need it below?

kbrock replied:
The old `preprocess_for_sql` modified the expression passed in as it worked. Sometimes that was wanted and other times it was not. So we needed to `exp.dup` in some of the operators and not in others. (this got too confusing for me)

Now, we are modifying the tree up front in `preprocess_exp!`. Mostly this is removing the token and reformatting `"5.megabytes"`. We get all the modifications out of the way up front. So the `reduce_exp` phase only removes nodes. And when it does that, it returns a different array (rather than modifying the array in place (i.e.: `exp.delete(operator)`).

I discovered this because the tests started failing in strange ways. You can see all the `freeze` statements I put in there to catch anything strange happening.

In the end, separating this into 2 phases allows us to fix the Demorgan bug and handle the differences between reducing ruby and sql.

jrafanie replied:
I think you're saying we don't need the `dup` 🤣

kbrock commented:
I started trying to track down whether the timezone is an issue.

The issue could show up if we reuse a miq expression across multiple calls - each containing a different timezone. (if we store the miq expression in a report, then it may be possible.

TODO: find ref to where we generate widgets based upon the timezones from an miq_group.


Now as for the prune_sql.

The issue could show up if we use a miq_expression in rbac and then use it not in rbac (and in that second usage, do not use the expression to filter the sql)

If we use it not in rbac first, then this will not be a problem but we will take a minor performance hit.

jrafanie commented:
@kbrock @Fryguy just confirming... we're not planning on backporting this to petrosian at this time, right?  I think it's a good direction going forward but would think it's a risk

jrafanie reviewed (approved):
LGTM, maybe we can merge in a bit, after we do the other backports for petrosian.  This touches several files so maybe it's good to wait a few days.

kbrock commented:
@jrafanie I will pull out the first commit with comments.
They touch a number of files. would be best to have this PR touch as few files as possible

jrafanie commented:
@Fryguy @kbrock what are your thoughts on this?  I think it's good but would probably like to merge and not yet backport it to petrosian so we can test it on master for a bit.

kbrock commented:
@jrafanie If you feel more comfortable, we can always have `@ruby` and `@ruby_only`. Either as 2 methods or the method using the param to determine the source.

Going forward, I would like to build upon this significantly https://github.com/ManageIQ/manageiq/issues/22397 (fry really wants to 🔥 `eval` and I don't blame him)

kbrock commented:
@Fryguy is now a good time to get this in?

kbrock commented:
update:
- added spec to ensure that `to_ruby` busted the cache when different arguments were passed in.

kbrock commented:
update:
- renamed `reduce_exp` to `prune_exp` since it prunes the expression according to the desired target

kbrock commented:
update:
- rebased to latest master
- fixed 2 rubocop whitespace issues

kbrock commented:
update:
- pulled busting cache into own commit

kbrock commented:
update:
- moved a change line from one commit to the busting cache commit (all worked, end result the same, but that commit would not have worked without this line change)
- rebased

kbrock commented:
Just ran into this today. This will speed up timelines quite a bit.

Fryguy commented:
LGTM - merging.
